codice/ddf
# [2.13.x] DDF-04275 Replace keyczar with tink

**NOTE: The second commit is a complete refactor, so the first commit should be ignored.**




#### What does this PR do?
Keyczar uses SHA1, which is insecure. Keyczar should be removed and replaced with google's newer crypto library, Tink.

This PR takes out Keyczar, replaces it with Tink, and abstracts the underlying cryptography logic into a `Crypter` class. The Tink packages, and Tink's necessary transient dependencies's packages, are added as extensions to bundle 0 (except for in `platform-osgi-configadmin` where the dependencies are embedded because of weird package resolution issues). The `Crypter` jar is embedded where it is needed.

#### Who is reviewing it? 
@brjeter 
@coyotesqrl
@garrettfreibott  
@stustison 

#### Select relevant component teams: 
@codice/security

#### How should this be tested?
1. Build, start, and standard install DDF. Verify all bundles come up without error.
1. Verify that ingesting and searching work.
1. Run `security:encrypt <any_value>` in the karaf console. Verify the returned result is encrypted and wrapped in `ENC()`.
1. Run `system:property solr.password` in the karaf console. Verify the value is encrypted and wrapped in `ENC()`.
1. In the distribution, navigate to `<DDF_HOME>/data/cache/org.eclipse.osgi/14/data/config/org/codice/ddf/catalog/ui/`, and verify that the encrypted values in the `config.config` file look similar to the encrypted values in [config-output.txt](https://github.com/codice/ddf/files/2838506/config-output.txt).
1. In the distribution, navigate to `<DDF_HOME>/etc/keysets/` and verify the directory contains 2 json files.
1. In the distribution, navigate to `<DDF_HOME>/etc/associatedData.properties` and verify there are 2 properties in the file.

#### What are the relevant tickets?
For GH Issues:
Fixes: #4275 

#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
